[DO NOT MERGE] Bug 1277304/1307741 - Migration-day changes for SCL3 only

edmorley opened the pull request:
On migration day, this branch will be deployed to SCL3 (but not to Heroku) to prepare for the switchover to Heroku. These changes will never be merged to master.

During the migration, the celery queues must reach zero, since unlike the MySQL database, anything left in the SCL3 rabbitmq instance will not be transferred to the Heroku one.

At the point at which the DB replication is shut off, we must also ensure no new DB writes are occurring in SCL3.

As such, this commit:
* Adds a dismissible maintenance overlay to the main Treeherder page, so users are informed of the downtime. The login buttons are also disabled, and made to re-show the maintenance info.
* Blocks any requests to the site that are not of type GET/HEAD/OPTIONS, from any user (whether Django or Hawk authenticated). This will prevent job submissions as well as user failure classifications/admin panel changes and so on.
* Disables the celery beat process, so new periodic tasks (such as pushlog/builds-4hr fetching, cycle_data etc) are not scheduled.
* Disables the pulse ingestion process, to stop pulling in new Taskcluster jobs from their Pulse queue. Jobs will safely build up in Pulse in the meantime.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/1627)
<!-- Reviewable:end -->

edmorley commented:
PR updated after bug 1211414, since the `treeherder-etl` Hawk ID no longer exists, and so doesn't need exempting from the API "down for maintenance" block.

edmorley commented:
Since we've now decided to do the Heroku migration midweek, we're going to make Treeherder read-only and just show a banner in the UI, rather than hide the UI behind a maintenance page. This is the banner I have so far (it appears on page load and can be dismissed in a manner similar to the keyboard shortcuts panel):

![screenshot](https://cloud.githubusercontent.com/assets/501702/16993498/9b987e1c-4e9b-11e6-93b1-900c5d435078.jpg)

@wlach @camd Thoughts? :-)

camd commented:
This seems like a great idea.  No reason not to let people look at the data that's already there.
Thanks, Ed!

edmorley commented:
PR updated with the following changes:
* Removed the maintenance page (that couldn't be bypassed) and replaced it with a maintenance information overlay (see screenshot in previous comment) that can be dismissed the same as the shortcuts help overlay.
* The permissions.py based approach of blocking requests that are making changes (ie anything but GET/HEAD/OPTIONS) has been replaced with a custom middleware approach, since the latter also protects against Django admin changes and is much simpler.
* I've added an explicit test for the HTTP 503 maintenance behaviour, and just skipped any other tests that fail due to not being able to submit, rather than make the PR hard to read with 100s of changes. (Even more tests otherwise needed fixing than in the last version of the PR, since the middleware approach is more complete at blocking).

A way to test:
* Run `./mange.py runserver` but with master checked out
* Open http://local.treeherder.mozilla.org
* Log into the local instance
* Only then update the Git checkout to this branch (and don't refresh the loaded UI)
* Then try eg retriggers and check that the UX is ok (this emulates eg a sheriff who has the UI open already when we do the deploy of this branch, and so won't initially see the maintenance banner)
* Then refresh the UI to see the maintenance banner

Limitations:
* Until bug 1290082 is fixed, not all UI actions that make requests will correctly display the full server maintenance message.
* The perf pages don't display the overlay, since they haven't been fitted with the keyboard shortcuts overlay. I don't think it's worth the effort hooking that up for the sake of a 30 min maintenance window, given the perfherder people are more advanced users who will inevitably already know about the maintenance, and who will likely already load/refresh the Treeherder homepage if they try making perfherder changes and requests fail.

edmorley commented:
Tested on stage, and checked:
* That New Relic graphs showed ingestion stopped
* That celery logs on the rabbitmq node showed the celerybeat and pulse listening processes were happily looping with the sleep/message cycle
* That the "client connections" list via MySQL Workbench showed no writes

edmorley commented:
PR rebased on master + the newly added `bin/run_read_pulse_resultsets` modified in the same manner as `bin/run_read_pulse_jobs`.

edmorley commented:
Updated with:
* Prod bug / newsgroup post links in the maintenance banner
* `retryable_task` countdown set to 5 seconds, not the exponential backup (otherwise queues will take ages to empty)

edmorley commented:
This has served it's purpose, so can now be closed out.
